File: src/main/java/org/academy/api/client/gui/widget/BlendQuadWidget.java
package org.academy.api.client.gui.widget;

import com.mojang.blaze3d.buffers.Std140Builder;
import com.mojang.blaze3d.buffers.Std140SizeCalculator;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.textures.FilterMode;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.DynamicUniformStorage;
import org.academy.api.client.Render;
import org.academy.api.client.Resource;
import org.academy.api.client.gui.command.ImageDrawCommand;
import org.academy.api.client.gui.command.PosTexRectDrawCommand;
import org.academy.api.client.gui.render.RenderContext;
import org.academy.api.client.render.UniformPayload;
import org.joml.Vector2f;
import org.joml.Vector4f;

import java.nio.ByteBuffer;
import java.util.List;

public class BlendQuadWidget extends AbstractWidget {
    public float marginTop = 4f;
    public float marginBottom = 4f;
    public float marginLeft = 4f;
    public float marginRight = 4f;
    public boolean drawLine = true;

    public float red;
    public float green;
    public float blue;

    @Override
    public void render(RenderContext context) {
        if (!isVisible()) return;

        var lp = getLayoutParams();
        var paddedWidth = getWidth() - lp.paddingLeft - lp.paddingRight;
        var paddedHeight = getHeight() - lp.paddingTop - lp.paddingBottom;

        // paddedHeight == 0 是预期行为喵
        if (paddedWidth <= 0 || paddedHeight < 0) return;

        var finalAlpha = getAlpha() * context.getAccumulatedAlpha();

        context.pose().pushPose();
        context.pose().translate(lp.paddingLeft, lp.paddingTop, 0);
        context.drawOrder().push();
        {
            // 极小值也需要渲染喵
            if (finalAlpha != 0) {
                var sdfCommand = new PosTexRectDrawCommand(
                        Render.RenderPipelines.SDF_SHARP_MARGIN,
                        paddedWidth,
                        paddedHeight,
                        0.0f,
                        0.0f,
                        1.0f,
                        1.0f,
                        List.of(),
                        List.of(new UniformPayload<>(
                                        "SdfUniforms",
                                SDFData.class, new SDFData(
                                                new Vector2f(paddedWidth, paddedHeight),
                                                new Vector4f(
                                                        marginLeft, marginTop, marginRight, marginBottom
                                                ),
                                                new Vector4f(red, green, blue, finalAlpha)
                                        ),
                                SDFData.UBO_SIZE
                                )
                        )
                );
                context.submit(sdfCommand);
            }

            if (drawLine) {
                context.drawOrder().advance();
                renderLines(context, context.getAccumulatedAlpha(), paddedWidth, paddedHeight);
            }
        }
        context.drawOrder().pop();
        context.pose().popPose();
    }

    private void renderLines(RenderContext context, float finalAlpha, float paddedWidth, float paddedHeight) {
        var sampler = RenderSystem.getSamplerCache().getClampToEdge(FilterMode.NEAREST);
        var textureManager = Minecraft.getInstance().getTextureManager();
        var lineTextureView = textureManager.getTexture(Resource.Textures.ELEMENT_LINE).getTextureView();
        var lineH = 4.0f;
        {
            var topLineCommand = new ImageDrawCommand(lineTextureView, sampler, paddedWidth, lineH, 0, 0, 1, 1, 1.0f, 1.0f, 1.0f, finalAlpha);
            context.submit(topLineCommand);
        }
        {
            context.pose().pushPose();
            context.pose().translate(0, Math.max(paddedHeight - lineH, 0), 0);
            var bottomLineCommand = new ImageDrawCommand(lineTextureView, sampler, paddedWidth, lineH, 0, 0, 1, 1, 1.0f, 1.0f, 1.0f, finalAlpha);
            context.submit(bottomLineCommand);
            context.pose().popPose();
        }
    }

    public record SDFData(Vector2f size, Vector4f margins,
                          Vector4f fillColor) implements DynamicUniformStorage.DynamicUniform {
        public static final int UBO_SIZE = new Std140SizeCalculator().putVec2().putVec4().putVec4().get();

        @Override
        public void write(ByteBuffer buffer) {
            Std140Builder.intoBuffer(buffer).putVec2(size).putVec4(margins).putVec4(fillColor);
        }
    }
}

File: src/main/kotlin/org/academy/api/client/gui/widget/BlendQuadWidget.kt
package org.academy.api.client.gui.widget

import com.mojang.blaze3d.buffers.Std140Builder
import com.mojang.blaze3d.buffers.Std140SizeCalculator
import com.mojang.blaze3d.systems.RenderSystem
import com.mojang.blaze3d.textures.FilterMode
import net.minecraft.client.renderer.DynamicUniformStorage.DynamicUniform
import org.academy.api.client.gui.command.ImageDrawCommand
import org.academy.api.client.gui.command.PosTexRectDrawCommand
import org.academy.api.client.gui.environment.UiEnvironment
import org.academy.api.client.gui.render.RenderContext
import org.academy.api.client.render.Render
import org.academy.api.client.render.UniformPayload
import org.academy.api.client.resources.R
import org.joml.Vector2f
import org.joml.Vector4f
import java.nio.ByteBuffer
import kotlin.math.max

class BlendQuadWidget : AbstractWidget() {
    var marginTop: Float = 4f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var marginBottom: Float = 4f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var marginLeft: Float = 4f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var marginRight: Float = 4f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var drawLine: Boolean = true
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }

    var red: Float = 0f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var green: Float = 0f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }
    var blue: Float = 0f
        set(value) {
            if (field != value) {
                field = value
                invalidate()
            }
        }

    override fun render(context: RenderContext) {
        if (!isVisible()) return

        val lp = layoutParams
        val paddedWidth = width - lp.paddingLeft - lp.paddingRight
        val paddedHeight = height - lp.paddingTop - lp.paddingBottom

        // paddedHeight == 0 是预期行为喵
        if (paddedWidth <= 0 || paddedHeight < 0) return

        val finalAlpha = alpha * context.accumulatedAlpha

        context.pose().pushPose()
        context.pose().translate(lp.paddingLeft, lp.paddingTop)
        context.drawOrder().push()
        run {
            // 极小值也需要渲染喵
            if (finalAlpha != 0f) {
                val sdfCommand = PosTexRectDrawCommand(
                    Render.RenderPipelines.SDF_SHARP_MARGIN,
                    paddedWidth,
                    paddedHeight,
                    0.0f,
                    0.0f,
                    1.0f,
                    1.0f,
                    listOf(),
                    listOf(
                        UniformPayload(
                            "SdfUniforms",
                            SDFData::class.java, SDFData(
                                Vector2f(paddedWidth, paddedHeight),
                                Vector4f(
                                    marginLeft, marginTop, marginRight, marginBottom
                                ),
                                Vector4f(red, green, blue, finalAlpha)
                            ),
                            SDFData.UBO_SIZE
                        )
                    )
                )
                context.submit(sdfCommand)
            }
            if (drawLine) {
                context.drawOrder().advance()
                renderLines(context, context.accumulatedAlpha, paddedWidth, paddedHeight)
            }
        }
        context.drawOrder().pop()
        context.pose().popPose()
    }

    private fun renderLines(context: RenderContext, finalAlpha: Float, paddedWidth: Float, paddedHeight: Float) {
        val sampler = RenderSystem.getSamplerCache().getClampToEdge(FilterMode.NEAREST)
        val lineTextureView = UiEnvironment.get().loadTexture(R.textures.gui.element.line)
        val lineH = 4.0f
        run {
            val topLineCommand = ImageDrawCommand(
                lineTextureView,
                sampler,
                paddedWidth,
                lineH,
                0f,
                0f,
                1f,
                1f,
                1.0f,
                1.0f,
                1.0f,
                finalAlpha
            )
            context.submit(topLineCommand)
        }
        run {
            context.pose().pushPose()
            context.pose().translate(0f, max(paddedHeight - lineH, 0f))
            val bottomLineCommand = ImageDrawCommand(
                lineTextureView,
                sampler,
                paddedWidth,
                lineH,
                0f,
                0f,
                1f,
                1f,
                1.0f,
                1.0f,
                1.0f,
                finalAlpha
            )
            context.submit(bottomLineCommand)
            context.pose().popPose()
        }
    }

    data class SDFData(
        val size: Vector2f, val margins: Vector4f,
        val fillColor: Vector4f
    ) : DynamicUniform {
        override fun write(buffer: ByteBuffer) {
            Std140Builder.intoBuffer(buffer).putVec2(size).putVec4(margins).putVec4(fillColor)
        }

        companion object {
            val UBO_SIZE: Int = Std140SizeCalculator().putVec2().putVec4().putVec4().get()
        }
    }
}
